File: agent-mesh-py/python/agent_mesh/amesh/verify.py
"""``amesh verify`` — verify a peer's GitHub binding.

Mirror of ``agent-mesh-cli/src/verify.rs``. Fetches
``https://github.com/<user>.keys`` (stdlib urllib, no extra deps) and
walks the response line-by-line until one ed25519 key validates the
binding, or we exhaust the list.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path
from urllib.error import URLError
from urllib.request import urlopen

from agent_mesh.core import GitHubBinding


async def run(binding_path: Path, github_user: str) -> int:
    """Fetch GitHub keys and try each ed25519 line against the binding.

    The fetch is sync (stdlib urllib) but the surrounding entry point
    is async so the CLI dispatcher can treat verify/announce/peers/
    listen/send uniformly via ``asyncio.run``.
    """
    if not binding_path.exists():
        print(
            f"error: read binding from {binding_path}: file not found",
            file=sys.stderr,
        )
        return 1
    try:
        binding_data = json.loads(binding_path.read_text())
    except (json.JSONDecodeError, OSError) as e:
        print(f"error: read binding from {binding_path}: {e}", file=sys.stderr)
        return 1
    try:
        binding = GitHubBinding.from_json(binding_data)
    except Exception as e:  # noqa: BLE001 — surface parse failures cleanly
        print(f"error: parse binding: {e}", file=sys.stderr)
        return 1

    url = f"https://github.com/{github_user}.keys"
    try:
        with urlopen(url, timeout=10) as resp:  # noqa: S310 — github keys are public
            keys_text = resp.read().decode("utf-8")
    except URLError as e:
        print(f"error: fetch {url}: {e.reason}", file=sys.stderr)
        return 1
    except Exception as e:  # noqa: BLE001 — last-ditch defensive
        print(f"error: fetch {url}: {e}", file=sys.stderr)
        return 1

    tried = 0
    for raw in keys_text.splitlines():
        line = raw.strip()
        if not line:
            continue
        # Only attempt ed25519 lines; the binding signature is ed25519
        # so anything else is structurally incompatible.
        if not line.startswith("ssh-ed25519 "):
            continue
        tried += 1
        if binding.try_verify_ssh_line(line):
            print("binding verified")
            print(f"  agent-mesh user: {binding.user_public().fingerprint().hex()}")
            print(f"  github user:     {github_user}")
            return 0

    print(
        f"error: no ed25519 key in github.com/{github_user}.keys "
        f"verified the binding (tried {tried} keys)",
        file=sys.stderr,
    )
    return 1

File: agent-mesh-cli/src/verify.rs
//! `amesh verify` — fetch a GitHub user's public keys and check that
//! one of them validates a binding file.

use agent_mesh_protocol::{ssh_pubkey_ed25519_bytes, GitHubBinding};
use anyhow::{Context, Result};
use ssh_key::PublicKey as SshPublicKey;
use std::path::PathBuf;

/// Verify a binding against `https://github.com/<github_user>.keys`.
///
/// Returns `Ok(())` on the first ed25519 key that validates the
/// binding; errors out if no key in the response succeeds.
pub async fn run(binding_path: PathBuf, github_user: String) -> Result<()> {
    let json = std::fs::read_to_string(&binding_path)
        .with_context(|| format!("read binding from {}", binding_path.display()))?;
    let binding: GitHubBinding = serde_json::from_str(&json)?;

    let url = format!("https://github.com/{github_user}.keys");
    let resp = reqwest::get(&url)
        .await
        .with_context(|| format!("fetch {url}"))?;
    if !resp.status().is_success() {
        anyhow::bail!(
            "github.com returned {} for {}.keys",
            resp.status(),
            github_user
        );
    }
    let keys_text = resp.text().await?;

    let mut tried = 0;
    for line in keys_text.lines() {
        if line.trim().is_empty() {
            continue;
        }
        let parsed = match SshPublicKey::from_openssh(line) {
            Ok(p) => p,
            Err(_) => continue,
        };
        let bytes = match ssh_pubkey_ed25519_bytes(&parsed) {
            Ok(b) => b,
            Err(_) => continue, // not an ed25519 key, skip
        };
        tried += 1;
        if binding.verify(&bytes).is_ok() {
            println!("binding verified");
            println!(
                "  agent-mesh user: {}",
                binding.user_pubkey.fingerprint().hex()
            );
            println!("  github user:     {github_user}");
            return Ok(());
        }
    }

    anyhow::bail!(
        "no ed25519 key in github.com/{github_user}.keys verified the binding (tried {tried} keys)"
    );
}
